### DISCUSSION TITLE
Have to run RustLsp debuggables twice using Lazy

### DISCUSSION THREAD

**Discussion by AddictArts** [mrcjkb/rustaceanvim#313] [Q&A]
Any help appreciated. Here is my config in the the table

```
{
 "mrcjkb/rustaceancim",
...
}
```

```
  config = function()
    -- Update this path
    local extension_path = vim.env.HOME .. "/.vscode/extensions/vadimcn.vscode-lldb-1.10.0/"
    local codelldb_path = extension_path .. "adapter/codelldb"
    local liblldb_path = extension_path .. "lldb/lib/liblldb"
    local this_os = vim.loop.os_uname().sysname

    -- The path is different on Windows
    if this_os:find "Windows_NT" then
      codelldb_path = extension_path .. "adapter\\codelldb.exe"
      liblldb_path = extension_path .. "lldb\\bin\\liblldb.dll"
    else
      -- The liblldb extension is .so for Linux and .dylib for MacOS
      liblldb_path = liblldb_path .. (this_os == "Linux" and ".so" or ".dylib")
    end

    local cfg = require "rustaceanvim.config"

    vim.g.rustaceanvim = {
      inlay_hints = {
        highlight = "NonText",
      },
      tools = {
        hover_actions = {
          auto_focus = true,
        },
      },
      server = {
        on_attach = function(client, bufnr)
          local il_hints = require "lsp-inlayhints"
          il_hints.on_attach(client, bufnr)
        end,
      },
      dap = {
        adapter = cfg.get_codelldb_adapter(codelldb_path, liblldb_path),
      },
    }
  end,
}
```

**Comment by mrcjkb** [comment]
Hey 👋

I provide a snippet for lazy.nvim in the readme: https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#installation

**Reply by AddictArts** [reply]
Yes indeed and that is exactly what I am doing; However, the DAP config section does not fit well with those lazy instructions. So, I adapted as shown above. I also had added the inlay support.

I suspect the issue is spinning up codelldb and the first time does that, but doesn't make the connection. The second time works as it is ready? maybe that is a possible reason.

I am new to this and new to lua. So, I adapted the advanced instructions to the lazy way and that is the config above. Thanks

**Reply by mrcjkb** [reply]
I understand. But why did you set `opts = {}`?
And what does `...` mean? Is there more, or just the `config` function?

What happens the first time you run `:RustLsp debugganles`? Do you get prompted to select a configuration? If not, then that just means rust-analyzer hasn't fully initialised yet (it can be slow to do so).
Do you always have to call it twice, or just the first time after starting up?

**Reply by AddictArts** [reply]
Apologies I was trying to just have the important parts, here is the entire file, no opts, that was from bolerplate.

```
return {
  "mrcjkb/rustaceanvim",
  version = "^4", -- Recommended
  dependencies = {
    {
      "lvimuser/lsp-inlayhints.nvim",
      opts = {},
    },
    "mfussenegger/nvim-dap",
    "jay-babu/mason-nvim-dap.nvim",
  },
  ft = { "rust" },
  keys = {
    {
      mode = "n",
      "<leader>rr",
      ":RustLsp runnables <CR>",
      desc = "RustLSP runnabes",
      ft = { "rust" },
    },
    {
      mode = "n",
      "<leader>rR",
      ":RustLsp runnables",
      desc = "RustLSP runnables [args[]]?",
      ft = { "rust" },
    },
    {
      mode = "n",
      "<leader>rx",
      ":RustLsp! runnables <CR>",
      desc = "RustLSP runnables!",
      ft = { "rust" },
    },
    {
      mode = "n",
      "<leader>rx",
      ":RustLsp! runnables <CR>",
      desc = "RustLSP runnables!",
      ft = { "rust" },
    },
  },
  config = function()
    -- Update this path
    local extension_path = vim.env.HOME .. "/.vscode/extensions/vadimcn.vscode-lldb-1.10.0/"
    local codelldb_path = extension_path .. "adapter/codelldb"
    local liblldb_path = extension_path .. "lldb/lib/liblldb"
    local this_os = vim.loop.os_uname().sysname

    -- The path is different on Windows
    if this_os:find "Windows_NT" then
      codelldb_path = extension_path .. "adapter\\codelldb.exe"
      liblldb_path = extension_path .. "lldb\\bin\\liblldb.dll"
    else
      -- The liblldb extension is .so for Linux and .dylib for MacOS
      liblldb_path = liblldb_path .. (this_os == "Linux" and ".so" or ".dylib")
    end

    local cfg = require "rustaceanvim.config"

    vim.g.rustaceanvim = {
      inlay_hints = {
        highlight = "NonText",
      },
      tools = {
        hover_actions = {
          auto_focus = true,
        },
      },
      server = {
        on_attach = function(client, bufnr)
          local il_hints = require "lsp-inlayhints"
          il_hints.on_attach(client, bufnr)
        end,
      },
      dap = {
        adapter = cfg.get_codelldb_adapter(codelldb_path, liblldb_path),
      },
    }
  end,
}
```

**Reply by mrcjkb** [reply]
- What happens the first time you run `:RustLsp debugganles`? Do you get prompted to select a configuration? If not, then that just means rust-analyzer hasn't fully initialised yet (it can be slow to do so).

- Do you always have to call it twice, or just the first time after starting up?

- Are you calling it directly or via a keymap? If the latter, does it behave the same when called directly?

**Reply by AddictArts** [reply]
Thanks. I get the prompt for bin or test. Then it says lldb is missing as I recall. Do it again and it works. Only the first time. I think it doesn’t get the codelldb setting the first time. Command calling it.

**Reply by mrcjkb** [reply]
When you get a chance, could you please post the exact message?
That sounds like it's either nvim-dap or your codelldb installation.

**Reply by AddictArts** [reply]
Sure, see the no adapter lldb, but I am using codelldb, which on the next run is found and works, hit breakpoints.

```
Debuggables
1: build --package clap-box --bin clap-box                                                                                                                                                  2: test --no-run --package clap-box --all-targets                                                                                                                                           Type number and <Enter> or click with the mouse (q or empty cancels): 1Compiling a debug build for debugging. This might take some time...
Press ENTER or type command to continue
No adapter exists named "lldb". See ":h dap-adapter" for more information
Press ENTER or type command to continue
```

**Reply by mrcjkb** [reply]
Does your project have a `.vscode/launch.json` file?
If it does, and it contains launch configurations for lldb, rustaceanvim will pick them up.

You can prevent this, by hard-coding the configuration.

```lua
vim.g.rustaceanvim = {
      dap = {
          configuration = {
                name = 'Rust debug client',
                type = 'codelldb',
                request = 'launch',
                stopOnEntry = false,
          },
      },
}
```

In the meantime, I will look into implementing a fix that doesn't load any launch configurations that conflict with the adapter setting.

**Reply by mrcjkb** [reply]
https://github.com/mrcjkb/rustaceanvim/commit/d6fd0b78e49ff4dd37070155e9f14fd26f2ef53f

**Reply by AddictArts** [reply]
Fixed, thank you. Would have filed it as an issue, but I thought it was a me issue. Very much appreciated, wow!
